fgpyo/sam/__init__.py

```diff
@@ -1036,55 +1040,96 @@ class ReadEditInfo:
-) -> ReadEditInfo:
+def calculate_edit_info(  # noqa: C901 (11 > 10)
+    rec: AlignedSegment,
+    reference_sequence: str,
+    n_as_match: bool = True,
+    reference_offset: Optional[int] = None,
+) -> Optional[ReadEditInfo]:
     """
     Constructs a `ReadEditInfo` instance giving summary stats about how the read aligns to the
-    reference.  Computes the number of mismatches, indels, indel bases and the SAM NM tag.
-    The read must be aligned.
+    reference.  Computes the number of mismatches, indels, indel bases as well as the
+    SAM NM and MD tags.
+
+    Calculation of NM and MD tags is based off of htsjdk:
+    https://github.com/samtools/htsjdk/blob/7034b33636b4cb9fec300a2136588e7c12c7ccd5/src/main/java/htsjdk/samtools/util/SequenceUtil.java#L964:L1029
+
+    Per the SAM specification (https://samtools.github.io/hts-specs/SAMtags.pdf), the NM tag
+    encapsulates the number of differences between the query read and reference sequence, counting
+    only A, C, G and T bases (case-insensitive). Everything else should be considered a mismatch
+    (e.g., ambiguity codes like R and N). We set the default of `n_as_match` to True to be
+    concordant with `htsjdk`, which treats an N->N as a match.
+
+    If the read is unmapped or the query sequence contains missing bases (`*`), returns None, as it
+    is not possible to recalculate the MD and NM tags without access to the query sequence and
+    reference sequence.
 
     Args:
         rec: the read/record for which to calculate values
-        reference_sequence: the reference sequence (or fragment thereof) that the read is
-            aligned to
+        reference_sequence: the reference sequence (or fragment thereof) to which the read is
+            aligned
+        n_as_match: if True, mirror htsjdk `calculateMdAndNmTags` and treat N->N as a match. If
+            False, N->N is treated as a mismatch. (default: True)
         reference_offset: if provided, assume that reference_sequence[reference_offset] is the
             first base aligned to in reference_sequence, otherwise use r.reference_start
 
     Returns:
         a ReadEditInfo with information about how the read differs from the reference
     """
-    assert not rec.is_unmapped, f"Cannot calculate edit info for unmapped read: {rec}"
+    if rec.is_unmapped or rec.query_sequence is None or rec.query_sequence == NO_QUERY_BASES:
+        return None
 
-    query_offset = 0
-    target_offset = reference_offset if reference_offset is not None else rec.reference_start
-    cigar = Cigar.from_cigartuples(rec.cigartuples)
+    query_offset: int = 0
+    target_offset: int = reference_offset if reference_offset is not None else rec.reference_start
+    cigar: Cigar = Cigar.from_cigartuples(rec.cigartuples)
 
     matches, mms, insertions, ins_bases, deletions, del_bases = 0, 0, 0, 0, 0, 0
-    ok_bases = {"A", "C", "G", "T"}
-
+    ok_bases: set[str] = {"A", "C", "G", "T"}
```

Comment:
issue:

To follow this sentence in the spec
> The special sequence base ‘=’ will always be considered to be a match, even if the reference is ambiguous at that point.

I think we need to add `=` to the ok bases so it will never be treated as a mismatch.

```suggestion
    ok_bases: set[str] = {"A", "C", "G", "T", "="}
```

We also need one guard added as a suggestion below.

Could you please add a unit test to test for `=` in the query bases?

Reply:
added: `test_calc_edit_info_with_equals_in_query`